### PR TITLE
MC dropout

### PR DESCRIPTION
**Pull request by markus-hinsche**
User story https://dev.azure.com/cgmorg/ChildGrowthMonitor/_workitems/edit/729

**Changes:**
- Calculate uncertainty score
- Visualize some depthmaps and there associated uncertainty score (see jupyter notebook)
- Visualize the uncertainty distribution for a subgroup (e.g. all scans tagged as 'good')
- Put into azure pipeline (Speed: [41.8 sec](https://ml.azure.com/experiments/id/4347c355-aa9d-4482-b7d9-2130b9aa7c44/runs/QA-pipeline_1609256537_fd692cd9?wsid=/subscriptions/9b82ecea-6780-4b85-8acf-d27d79028f07/resourcegroups/cgm-ml-prod/workspaces/cgm-azureml-prod&tid=006dabd7-456d-465b-a87f-f7d557e319c8#outputsAndLogs) for 664 artifacts/20scans given 16predictions per artifact) (prediction without uncertainty: 13.1 sec) (randomly sample one artifact for each scan)
- Calculate the mean uncertainty of scans of type standing-front only

**Result:**
_Visualization of some depthmaps and there associated uncertainty score_
image

_Visualization of the uncertainty distribution_
- On the left: all scans
- On the right: only standing-front
image
dropout strength =1 NUM_PREDICTIONS=16[Run](https://dev.azure.com/cgmorg/ChildGrowthMonitor/_build/results?buildId=2246&view=artifacts&pathAsName=false&type=publishedArtifacts)

image
dropout strength =2 NUM_PREDICTIONS=16[Run](https://dev.azure.com/cgmorg/ChildGrowthMonitor/_build/results?buildId=2247&view=artifacts&pathAsName=false&type=publishedArtifacts)

image
dropout strength =0.5 NUM_PREDICTIONS=16[Run](https://dev.azure.com/cgmorg/ChildGrowthMonitor/_build/results?buildId=2248&view=artifacts&pathAsName=false&type=publishedArtifacts)

image
dropout strength =0.1 NUM_PREDICTIONS=16 [Run](https://dev.azure.com/cgmorg/ChildGrowthMonitor/_build/results?buildId=2249&view=artifacts&pathAsName=false&type=publishedArtifacts)

image
dropout strength=1, NUM_PREDICTIONS=64 [Run](https://dev.azure.com/cgmorg/ChildGrowthMonitor/_build/results?buildId=2250&view=artifacts&pathAsName=false&type=publishedArtifacts)

**Findings:**
- maybe NUM_PREDICTIONS=16 is enough
- The dropout strengths we tried performed similarly

**Duration:**
- for 200scans
    - Total time for prediction experiment: 32.5 sec
    - Total time for uncertainty prediction experiment: 38.0 sec
- approximally doubles the prediction time (NUM_PREDICTIONS = 16)

**Potential future work:**
- uncertainty only 1 age group
- what is the metric (one number) for this uncertainty

Before merging: Revert `pr: none`

### CODE REVIEW

**Comment by markus-hinsche** [comment]
Note: we have these dropout layers:
```
        model.add(layers.Dropout(0.05))
        model.add(layers.Dropout(0.075))
        model.add(layers.Dropout(0.1))
        model.add(layers.Dropout(0.125))
        model.add(layers.Dropout(0.15))
        model.add(layers.Dropout(0.175))
        model.add(layers.Dropout(0.2))
        model.add(layers.Dropout(0.25))
```

Note: we have some experiments with and some without dropout. Examples:
- Run3: dropout false
- Run17: dropout True

File: src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py [hunk 1]
```diff
@@ -53,16 +63,79 @@ def py_load_pickle(path, max_value):
...
+    Returns:
+        The standard deviation of multiple predictions
+    """
+    one_batch = np.repeat(X, NUM_PREDICTIONS, axis=0)
+    predictions = model(one_batch, training=True)
+    std = tf.math.reduce_std(predictions)
+    return std
+
+
+def change_dropout_strength(model: tf.keras.Model, dropout_strength: float) -> tf.keras.Model:
```
Comment (gangooteli) [inline src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py:93]: Suggestion: In future, this utility can be moved into model utils.

File: src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py [hunk 2]
```diff
@@ -214,5 +286,27 @@ def get_prediction(model_path, dataset_evaluation):
...
+        assert COLUMN_NAME_GOODBAD in df
+
+        # Sample one artifact per scan (qrcode, scantype combination)
+        df_sample = df.groupby(['qrcode', 'scantype']).apply(lambda x: x.sample(1))
+
+        # Prepare uncertainty prediction on these artifacts
+        dataset_sample = prepare_sample_dataset(df_sample, dataset_path)
+
+        # Predict uncertainty
+        uncertainty_list_one = get_prediction_uncertainty(model_path, dataset_sample)
```
Comment (gangooteli) [inline src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py:300]: Please rename the uncertainty_list_one to uncertainty_list if there is no specific reason to use one as suffix.

`uncertainty_list = get_prediction_uncertainty(model_path, dataset_sample) `

File: src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py [hunk 3]
```diff
@@ -1,40 +1,50 @@
...
-qa_config = import_module(args.qa_config_module)
 MODEL_CONFIG = qa_config.MODEL_CONFIG
 EVAL_CONFIG = qa_config.EVAL_CONFIG
 DATA_CONFIG = qa_config.DATA_CONFIG
 RESULT_CONFIG = qa_config.RESULT_CONFIG
 
 RUN_ID = MODEL_CONFIG.RUN_ID
 
+NUM_PREDICTIONS = 16
+DROPOUT_STRENGTH = 1  # 1.0 means like original model
```
Comment (gangooteli) [inline src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py:46]: Can these two be part of config file?

File: src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py [hunk 4]
```diff
@@ -53,16 +63,79 @@ def py_load_pickle(path, max_value):
...
+    return std
+
+
+def change_dropout_strength(model: tf.keras.Model, dropout_strength: float) -> tf.keras.Model:
+    """Duplicate a model while adjusting the dropout rate"""
+    new_model = Sequential(name="new_model")
+    for layer_ in model.layers:
+        layer = copy.copy(layer_)
+        if isinstance(layer, layers.core.Dropout):
+            layer.rate = min(0.999, layer.rate * dropout_strength)
```
Comment (gangooteli) [inline src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py:99]: Not able to understand the line. Could you explain this?

Comment (markus-hinsche) [reply src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py:99]: This will set the dropout rate a ratio from range [0.0, 1.0] according to DROPOUT_STRENGTH. If DROPOUT_STRENGTH is set so very large that it would exceed 1.0, then we set it to 0.999 instead.

Comment (markus-hinsche) [reply src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py]: I also added a comment

**Review by gangooteli — APPROVED** [review]
